lib/hanami/application.rb

```diff
@@ -1,64 +1,282 @@
+        container.start(*args)
+      end
+
+      def keys
+        container.keys
+      end
+
+      def [](*args)
+        container.[](*args)
+      end
+
+      def resolve(*args)
+        container.resolve(*args)
+      end
+
+      def boot(&block)
+        return self if booted?
+
+        init
+
+        container.configure do; end # force after configure hook
 
-      def routes(&blk)
+        container.finalize!(&block)
+
+        slices.each do |slice|
+          slice.boot
+        end
+
+        @booted = true
+        self
+      end
+
+      def routes(&block)
         @_mutex.synchronize do
-          if blk.nil?
-            raise "Hanami.application_class.routes not configured" unless defined?(@_routes)
+          if block.nil?
+            raise "Hanami.application.routes not configured" unless defined?(@_routes)
 
             @_routes
           else
-            @_routes = Routes.new(&blk)
+            @_routes = block
           end
         end
       end
+
+      MODULE_DELIMITER = "::"
+      private_constant :MODULE_DELIMITER
+
+      def application_module
+        inflector.constantize(name.split(MODULE_DELIMITER)[0..-2].join(MODULE_DELIMITER))
+      end
+
+      def application_name
+        inflector.underscore(application_module.to_s)
+      end
+
+      def root
+        # TODO: do we need anything more sophisticated than this? This is how
+        # Dry::System::Container determines its root by default, anyway.
+        Dir.pwd
+      end
+
+      def inflector
+        # TODO: might be good if we provided access as `config.inflector` too
+        config.inflections
+      end
+
+      private
+
+      def inited?
+        !!@inited
+      end
+
+      def booted?
+        !!@booted
+      end
+
+      def prepare_container
+        define_container.tap do |container|
+          configure_container container
+        end
+      end
+
+      def prepare_deps_module
+        define_deps_module
+      end
+
+      def define_container
+        begin
+          require "#{application_name}/container"
+          application_module.const_get :Container
+        rescue LoadError, NameError
+          Class.new(Dry::System::Container).tap do |container|
+            application_module.const_set :Container, container
+          end
+        end
+      end
+
+      def configure_container(container)
+        container.use :env, inferrer: -> { Hanami.env }
+        container.use :notifications
+
+        container.configure do |config|
+          config.auto_register = "lib/#{application_name}" # TODO: get from config somehow?
+          config.default_namespace = application_name
+        end
+
+        container.load_paths! "lib"
+
+        unless container.key?(:inflector)
+          container.register :inflector, inflector
+        end
+
+        unless container.key?(:logger)
+          require "hanami/logger"
+          container.register :logger, Hanami::Logger.new(configuration.logger)
+        end
```

Comment:
Would checking `key?(:logger)` result in lazy-booting `:logger`? IIRC I had an issue with it, I had a logger bootable component yet the default logger would kick in regardless.

Reply:
@solnic Hmmm... if dry-system isn't already lazy-loading a component when checked by `key?`, then it _should_ be. I'll a TODO to make sure we follow up on this.